src/libPMacc/include/cuSTL/algorithm/kernel/run-time/Foreach.hpp

```diff
@@ -52,6 +53,28 @@ namespace kernel
 namespace RT
 {
 
+/** Return a suitable cuda blockDim for a given gridDim.
+ *
+ * @param gridDim 1D, 2D or 3D grid size
+ * @return cuda blockDim
+ */
+template<int dim>
+math::Size_t<3> getBestCudaBlockDim(math::Size_t<dim> gridDim)
+{
+    math::Size_t<3> result = math::Size_t<3>::create(1);
+
+    /* The greatest common divisor of each component of the volume size
+     * and a certain power of two value gives the best suitable block size */
+    boost::math::gcd_evaluator<size_t> gcd; /* greatest common divisor */
+    size_t maxValues[] = {16, 16, 4}; /* maximum values for each axis */
```

Comment:
can we define this a bit more central and add a reasoning why we don't want to use more than 1024 threads? (e.g., what does sm_20+ allow)